Mathlib/Data/Nondet/Basic.lean

```diff
@@ -0,0 +1,176 @@
+(The backtrackable state in each will be identical:
+whatever the state was when we first read from the result.)
+-/
+def ofList [Alternative m] (L : List α) : Nondet m α := ofListM (L.map pure)
+
+/--
+Squash a monadic nondeterministic value to a nondeterministic value.
+-/
+def squash (L : Unit → m (Nondet m α)) : Nondet m α := MLList.squash L
+
+/-- Apply a function which returns values in the monad to every alternative of a `Nondet m α`. -/
+partial def mapM (f : α → m β) (L : Nondet m α) : Nondet m β :=
+  L.casesM
+    (fun _ => return .nil)
+    (fun ⟨s, x⟩ xs => do
+      restoreState s
+      let r ← f x
+      let s' ← saveState
+      return .cons (s', r) (Nondet.mapM f xs))
+
+/-- Apply a function to each alternative in a `Nondet m α` . -/
+def map (f : α → β) (L : Nondet m α) : Nondet m β :=
+  L.mapM fun a => pure (f a)
+
+/-- Filter a nondeterministic value using a monadic predicate. -/
+partial def filterM (p : α → m (ULift Bool)) (L : Nondet m α) : Nondet m α :=
+  L.casesM
+    (fun _ => return .nil)
+    (fun ⟨s, x⟩ xs => do
+      restoreState s
+      let r := (← p x).down
+      let s' ← saveState
+      return if r then do
+        .cons (s', x) (filterM p xs)
+      else
+        (filterM p xs))
+
+/-- Filter a nondeterministic value. -/
+def filter (p : α → Bool) (L : Nondet m α) : Nondet m α :=
+  L.filterM fun a => pure <| .up (p a)
+
+/-- Filter and map a nondeterministic value using a monadic function which may return `none`. -/
+partial def filterMapM (f : α → m (Option β)) (L : Nondet m α) : Nondet m β :=
+  L.casesM
+    (fun _ => return .nil)
+    (fun ⟨s, x⟩ xs => do
+      restoreState s
+      match ← f x with
+      | some b => do return .cons (← saveState, b) (filterMapM f xs)
+      | none => return filterMapM f xs)
+
+/-- Filter and map a nondeterministic value. -/
+def filterMap (f : α → Option β) (L : Nondet m α) : Nondet m β :=
+  L.filterMapM fun a => pure <| (f a)
+
+/--
+Find the first alternative in a nondeterministic value, as a monadic value.
+-/
+def head [Alternative m] (L : Nondet m α) : m α := (·.2) <$> MLList.head L
+
+/--
+Find the value of a monadic function on the first alternative in a nondeterministic value
+where the function succeeds.
+-/
+def firstM [Alternative m] (L : Nondet m α) (f : α → m (Option β)) : m β :=
+  MLList.firstM L (fun (s, a) => do restoreState s; f a)
+
+/--
+Convert a non-deterministic value into a lazy list, keeping the backtrackable state.
+Be careful that monadic operations on the `MLList` will not respect this state!
+-/
+def toMLList' (L : Nondet m α) : MLList m (σ × α) := L
+
+/--
+Convert a non-deterministic value into a lazy list, by discarding the backtrackable state.
+-/
+def toMLList (L : Nondet m α) : MLList m α := MLList.map (fun (_, a) => a) L
+
+end Nondet
+
+/-!
+As parts of the general `MLList` API may modify state incorrectly, we mark `Nondet` irreducible
+here to prevent accidental usage of the rest of the API.
+-/
+attribute [irreducible] Nondet
```

Comment:
How annoying is it to make `Nondet` a structure instead? If you do that, then you get `toMLList'` for free, and don't need to use irreducibility.

Reply:
If it's not too hard, I think making `Nondet` be a one-filed structure is a great idea. I'm finding it hard to verify that the correct monad operations are being applied throughout this code... but this would help with making things be more explicit.

Reply:
This was a great suggestion. It is all shorter, and more correct, now. As I've completely rewritten the file, I'm dismissing many of the comments below.